pimd/pim6_mld.c

```diff
@@ -0,0 +1,3011 @@
+			     const struct gm_subscriber *b)
+{
+	return IPV6_ADDR_CMP(&a->addr, &b->addr);
+}
+
+static uint32_t gm_subscriber_hash(const struct gm_subscriber *a)
+{
+	return jhash(&a->addr, sizeof(a->addr), 0xd0e94ad4);
+}
+
+DECLARE_HASH(gm_subscribers, struct gm_subscriber, itm, gm_subscriber_cmp,
+	     gm_subscriber_hash);
+
+static struct gm_subscriber *gm_subscriber_findref(struct gm_if *gm_ifp,
+						   pim_addr addr)
+{
+	struct gm_subscriber ref = {}, *ret;
+
+	ref.addr = addr;
+	ret = gm_subscribers_find(gm_ifp->subscribers, &ref);
+	if (ret)
+		ret->refcount++;
+	return ret;
+}
+
+static struct gm_subscriber *gm_subscriber_get(struct gm_if *gm_ifp,
+					       pim_addr addr)
+{
+	struct gm_subscriber ref = {}, *ret;
+
+	ref.addr = addr;
+	ret = gm_subscribers_find(gm_ifp->subscribers, &ref);
+
+	if (!ret) {
+		ret = XCALLOC(MTYPE_GM_SUBSCRIBER, sizeof(*ret));
+		ret->iface = gm_ifp;
+		ret->addr = addr;
+		ret->refcount = 1;
+		monotime(&ret->created);
+		gm_packets_init(ret->packets);
+
+		gm_subscribers_add(gm_ifp->subscribers, ret);
+	}
+	return ret;
+}
+
+static void gm_subscriber_drop(struct gm_subscriber **subp)
+{
+	struct gm_subscriber *sub = *subp;
+	struct gm_if *gm_ifp;
+
+	if (!sub)
+		return;
+	gm_ifp = sub->iface;
+
+	*subp = NULL;
+	sub->refcount--;
+
+	if (sub->refcount)
+		return;
+
+	gm_subscribers_del(gm_ifp->subscribers, sub);
+	XFREE(MTYPE_GM_SUBSCRIBER, sub);
+}
+
+/****************************************************************************/
+
+/* bundle query timer values for combined v1/v2 handling */
+struct gm_query_timers {
+	unsigned int qrv;
+	unsigned int max_resp_ms;
+	unsigned int qqic_ms;
+
+	struct timeval fuzz;
+	struct timeval expire_wait;
+};
+
+static void gm_expiry_calc(struct gm_query_timers *timers)
+{
+	unsigned int expire =
+		(timers->qrv - 1) * timers->qqic_ms + timers->max_resp_ms;
+	ldiv_t exp_div = ldiv(expire, 1000);
+
+	timers->expire_wait.tv_sec = exp_div.quot;
+	timers->expire_wait.tv_usec = exp_div.rem * 1000;
+	timeradd(&timers->expire_wait, &timers->fuzz, &timers->expire_wait);
+}
+
+static void gm_sg_free(struct gm_sg *sg)
+{
+	/* t_sg_expiry is handled before this is reached */
+	THREAD_OFF(sg->t_sg_query);
+	gm_packet_sg_subs_fini(sg->subs_negative);
+	gm_packet_sg_subs_fini(sg->subs_positive);
+	XFREE(MTYPE_GM_SG, sg);
+}
+
+/* clang-format off */
+static const char *const gm_states[] = {
+	[GM_SG_NOINFO]			= "NOINFO",
+	[GM_SG_JOIN]			= "JOIN",
+	[GM_SG_JOIN_EXPIRING]		= "JOIN_EXPIRING",
+	[GM_SG_PRUNE]			= "PRUNE",
+	[GM_SG_NOPRUNE]			= "NOPRUNE",
+	[GM_SG_NOPRUNE_EXPIRING]	= "NOPRUNE_EXPIRING",
+};
+/* clang-format on */
+
+CPP_NOTICE("TODO: S,G entries in EXCLUDE (i.e. prune) unsupported");
+/* tib_sg_gm_prune() below is an "un-join", it doesn't prune S,G when *,G is
+ * joined.  Whether we actually want/need to support this is a separate
+ * question - it is almost never used.  In fact this is exactly what RFC5790
```

Comment:
Are we going to be complaint with RFC 3810 or RFC 5790.  Are there plans to support S,G Exclude

Reply:
S,G exclude is supported in MLD but not propagated into PIM. I have not decided yet whether it is worth the effort to implement full S,G Exclude support.

Reply:
Without being propagated to PIM, the host will still receive the excluded source traffic, which defeats the purpose of SG exclude report.

Reply:
@plsaranya yes it does. I have implemented MLD to be capable of supporting S,G Exclude because it would have been difficult to add it later. But S,G Exclude is a very high cost of code complexity for ultimately non-existing use cases. It is extremely questionable whether completing the support makes sense.

If anything, the question at this point is whether anyone can provide a good use case that can "pay" for the long-term complexity and maintenance cost. The IETF recognized that S,G exclude is essentially never used and therefore created RFC 5790. We can make the same decision and go with RFC 5790.

Reply:
@eqvinox  I am fine with RFC5790. Since MLD has SG exclude support but PIM not, I had this question. Can we make it completely with RFC 5790 then